Content.Server/_Moffstation/GameTicking/Rules/GunGameRuleSystem.cs

```diff
@@ -0,0 +1,229 @@
+    }
+
+    /// <summary>
+    /// Deletes a GunGameRewardTrackerComponent's gear.
+    /// </summary>
+    /// <param name="gear">The component with the gear.</param>
+    /// <remarks>
+    /// Right now this is mostly just a foreach loop which deletes all the entities in the list.
+    /// This does trigger an exception in the client if a gun is deleting while it's bullets are still active in the world.
+    /// </remarks>
+    private void DeleteCurrentWeapons(GunGameRewardTrackerComponent gear)
+    {
+        foreach (var entity in gear.CurrentRewards)
+        {
+            QueueDel(entity);
+        }
+    }
+
+    /// <summary>
+    /// Progresses the specified user's reward queue.
+    /// If their queue is empty then they've won so we trigger the round end.
+    /// </summary>
+    /// <param name="userId">The player's user ID</param>
+    /// <param name="rule">The GunGameRuleComponent for this gamerule</param>
+    private void ProgressPlayerReward(NetUserId userId, GunGameRuleComponent rule)
+    {
+        if (!rule.PlayerRewards.TryGetValue(userId, out var rewardQueue))
+            return;
+
+
+        if (rewardQueue.Count > 1)
+        {
+            rewardQueue.Dequeue();
+            return;
+        }
+
+        rule.Victor = userId;
+        _roundEnd.EndRound(rule.RestartDelay);
+    }
+
+    /// <summary>
+    /// Spawns the player's current weapon in their queue, intended to be used on spawn and when they upgrade weapons.
+    /// </summary>
+    /// <param name="userId">The player's user ID</param>
+    /// <param name="rule">The GunGameRuleComponent for this gamerule</param>
+    private void SpawnCurrentWeapon(NetUserId userId, GunGameRuleComponent rule)
+    {
+        if (!_mind.TryGetMind(userId, out var _, out var mind))
+            return;
+
+        if (mind.OwnedEntity is not { } playerEntity)
+            return;
+
+        var gear = EnsureComp<GunGameRewardTrackerComponent>(playerEntity);
+        DeleteCurrentWeapons(gear);
+
+        if (!rule.PlayerRewards.TryGetValue(userId, out var gearQueue))
+        {
+            gearQueue = new Queue<EntProtoId>(rule.RewardSpawnsQueue);
+            rule.PlayerRewards.Add(userId, gearQueue);
+        }
+
+        if (gearQueue.Count == 0) // If we somehow try to spawn somebody's weapon after they win
+            return;
+
+        // Peek the player's queue, and spawn the item there
+        var newWeapon = Spawn(gearQueue.Peek());
+
+        // Put it in their hands
+        _hands.TryForcePickupAnyHand(playerEntity, newWeapon);
+        gear.CurrentRewards.Add(newWeapon);
+    }
+
+    /// <summary>
+    /// Formats the scoreboard for the end of round screen.
+    /// </summary>
+    /// <param name="uid">The gamemode uid</param>
+    /// <param name="component">The GunGameRuleComponent for this gamemode</param>
+    /// <returns></returns>
+    private FormattedMessage GetScoreboard(EntityUid uid, GunGameRuleComponent? component = null)
```

Comment:
```suggestion
    private FormattedMessage GetScoreboard(Entity<GunGameRuleComponent?> entity)
```
And then you can `Resolve` straight into a `ref entity.Comp` and it just works, or so I have been lead to believe.